File: analysis/calendar_analyzer.py
#!/usr/bin/env python3
"""
Calendar Screenshot Analyzer for "Smash Your Week" Game

This module analyzes calendar screenshots (Google Calendar, Outlook, etc.)
and extracts event blocks with their positions, colors, and text.

Output: Structured JSON ready for game generation.
"""

import cv2
import numpy as np
import pytesseract
from PIL import Image
import json
import argparse
from pathlib import Path
from dataclasses import dataclass, asdict
from typing import List, Tuple, Optional
import colorsys


@dataclass
class CalendarEvent:
    """Represents a detected calendar event block."""
    x: float  # Relative X position (0-1)
    y: float  # Relative Y position (0-1)
    width: float  # Relative width (0-1)
    height: float  # Relative height (0-1)
    color: str  # Hex color code
    text: str  # Event text/title
    confidence: float  # Detection confidence (0-1)


@dataclass
class CalendarGrid:
    """Represents the detected calendar grid structure."""
    days: List[str]
    time_range: dict
    columns: int
    rows: int


@dataclass
class AnalysisResult:
    """Complete analysis result."""
    image_width: int
    image_height: int
    calendar: dict
    events: List[dict]


class CalendarAnalyzer:
    """
    Analyzes calendar screenshots to extract event blocks.

    Uses OpenCV for image processing and Tesseract for OCR.
    Designed for Google Calendar, Outlook, and similar calendar layouts.
    """

    # Common calendar background colors (in HSV ranges)
    BACKGROUND_COLORS = [
        # White/light gray backgrounds
        ((0, 0, 200), (180, 30, 255)),
        # Very light colors
        ((0, 0, 240), (180, 20, 255)),
    ]

    # Minimum block size relative to image (filter out noise)
    MIN_BLOCK_WIDTH_RATIO = 0.02
    MIN_BLOCK_HEIGHT_RATIO = 0.01
    MAX_BLOCK_WIDTH_RATIO = 0.5
    MAX_BLOCK_HEIGHT_RATIO = 0.3

    def __init__(self, debug: bool = False):
        self.debug = debug
        self.debug_images = {}

    def analyze(self, image_path: str) -> AnalysisResult:
        """
        Main entry point: analyze a calendar screenshot.

        Args:
            image_path: Path to the calendar screenshot

        Returns:
            AnalysisResult with detected events and grid structure
        """
        # Load image
        image = cv2.imread(image_path)
        if image is None:
            raise ValueError(f"Could not load image: {image_path}")

        height, width = image.shape[:2]

        # Detect calendar grid structure
        grid = self._detect_grid(image)

        # Detect event blocks
        events = self._detect_events(image)

        # Enhance text detection with OCR
        events = self._extract_text(image, events)

        # Convert to relative coordinates
        events = self._normalize_coordinates(events, width, height)

        # Filter and validate events
        events = self._filter_events(events)

        return AnalysisResult(
            image_width=width,
            image_height=height,
            calendar={
                "days": grid.days,
                "timeRange": grid.time_range,
                "columns": grid.columns,
                "rows": grid.rows
            },
            events=[asdict(e) for e in events]
        )

    def _detect_grid(self, image: np.ndarray) -> CalendarGrid:
        """
        Detect the calendar grid structure (days, time slots).

        Uses line detection and header analysis.
        """
        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        height, width = image.shape[:2]

        # Detect vertical lines (day separators)
        edges = cv2.Canny(gray, 50, 150, apertureSize=3)
        lines = cv2.HoughLinesP(edges, 1, np.pi/180, threshold=100,
                                minLineLength=height*0.3, maxLineGap=10)

        vertical_lines = []
        if lines is not None:
            for line in lines:
                x1, y1, x2, y2 = line[0]
                # Check if line is mostly vertical
                if abs(x1 - x2) < 10:
                    vertical_lines.append((x1 + x2) // 2)

        # Estimate number of columns (days)
        vertical_lines = sorted(set(vertical_lines))
        columns = max(1, len(vertical_lines) + 1) if vertical_lines else 7

        # Try to detect day names from header
        days = self._detect_day_headers(image)
        if not days:
            days = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"][:columns]

        # Try to detect time range from left side
        time_range = self._detect_time_range(image)

        # Estimate rows based on typical calendar layout
        rows = 24  # Assume hourly slots

        return CalendarGrid(
            days=days,
            time_range=time_range,
            columns=min(columns, 7),
            rows=rows
        )

    def _detect_day_headers(self, image: np.ndarray) -> List[str]:
        """Try to detect day names from the calendar header."""
        height, width = image.shape[:2]

        # Look at top portion of image for headers
        header_region = image[0:int(height*0.1), :]

        # OCR on header
        try:
            text = pytesseract.image_to_string(header_region)
            # Look for day patterns
            days = []
            day_patterns = [
                "monday", "tuesday", "wednesday", "thursday",
                "friday", "saturday", "sunday",
                "mon", "tue", "wed", "thu", "fri", "sat", "sun",
                "lun", "mar", "mer", "jeu", "ven", "sam", "dim"  # French
            ]
            text_lower = text.lower()
            for pattern in day_patterns:
                if pattern in text_lower:
                    days.append(pattern.capitalize()[:3])
            return days[:7]
        except Exception:
            return []

    def _detect_time_range(self, image: np.ndarray) -> dict:
        """Try to detect time range from left margin."""
        height, width = image.shape[:2]

        # Look at left portion for time labels
        time_region = image[:, 0:int(width*0.1)]

        try:
            text = pytesseract.image_to_string(time_region)
            # Look for time patterns
            import re
            times = re.findall(r'(\d{1,2})[:\.]?(\d{2})?\s*(am|pm)?', text.lower())

            if times:
                hours = [int(t[0]) for t in times]
                if hours:
                    return {
                        "start": f"{min(hours):02d}:00",
                        "end": f"{max(hours):02d}:00"
                    }
        except Exception:
            pass

        # Default time range
        return {"start": "08:00", "end": "18:00"}

    def _detect_events(self, image: np.ndarray) -> List[CalendarEvent]:
        """
        Detect calendar event blocks using color segmentation.

        Strategy:
        1. Convert to HSV for better color separation
        2. Find colored regions (non-white/gray)
        3. Use contour detection to find block boundaries
        4. Extract dominant color for each block
        """
        height, width = image.shape[:2]
        hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)

        events = []

        # Method 1: Find saturated colored regions (event blocks usually have color)
        saturation = hsv[:, :, 1]
        value = hsv[:, :, 2]

        # Create mask for colored regions (high saturation or specific patterns)
        colored_mask = (saturation > 40) | ((saturation > 20) & (value < 200))
        colored_mask = colored_mask.astype(np.uint8) * 255

        # Also detect regions that differ from white background
        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        _, diff_mask = cv2.threshold(gray, 245, 255, cv2.THRESH_BINARY_INV)

        # Combine masks
        combined_mask = cv2.bitwise_or(colored_mask, diff_mask)

        # Clean up mask
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (5, 5))
        combined_mask = cv2.morphologyEx(combined_mask, cv2.MORPH_CLOSE, kernel)
        combined_mask = cv2.morphologyEx(combined_mask, cv2.MORPH_OPEN, kernel)

        if self.debug:
            self.debug_images['mask'] = combined_mask

        # Find contours
        contours, _ = cv2.findContours(combined_mask, cv2.RETR_EXTERNAL,
                                       cv2.CHAIN_APPROX_SIMPLE)

        # Process each contour
        for contour in contours:
            x, y, w, h = cv2.boundingRect(contour)

            # Filter by size (must be reasonable block size)
            if (w < width * self.MIN_BLOCK_WIDTH_RATIO or
                h < height * self.MIN_BLOCK_HEIGHT_RATIO or
                w > width * self.MAX_BLOCK_WIDTH_RATIO or
                h > height * self.MAX_BLOCK_HEIGHT_RATIO):
                continue

            # Filter by aspect ratio (blocks are usually wider than tall or square-ish)
            aspect_ratio = w / h if h > 0 else 0
            if aspect_ratio < 0.3 or aspect_ratio > 20:
                continue

            # Extract dominant color
            block_region = image[y:y+h, x:x+w]
            color = self._get_dominant_color(block_region)

            # Skip if color is too close to white/gray (likely background)
            if self._is_background_color(color):
                continue

            # Calculate confidence based on color saturation and contour area
            area_ratio = cv2.contourArea(contour) / (w * h) if w * h > 0 else 0
            confidence = min(1.0, area_ratio * 1.5)

            events.append(CalendarEvent(
                x=float(x),
                y=float(y),
                width=float(w),
                height=float(h),
                color=self._rgb_to_hex(color),
                text="",
                confidence=confidence
            ))

        return events

    def _get_dominant_color(self, region: np.ndarray) -> Tuple[int, int, int]:
        """
        Get the dominant color of a region.

        Uses k-means clustering to find the most common color,
        excluding very light colors (likely text or background).
        """
        if region.size == 0:
            return (128, 128, 128)

        # Reshape to list of pixels
        pixels = region.reshape(-1, 3).astype(np.float32)

        # Filter out very light pixels (likely background/text)
        hsv_pixels = cv2.cvtColor(region, cv2.COLOR_BGR2HSV).reshape(-1, 3)
        colored_mask = (hsv_pixels[:, 1] > 30) | (hsv_pixels[:, 2] < 200)
        colored_pixels = pixels[colored_mask]

        if len(colored_pixels) < 10:
            # Fallback: use mean of all pixels
            return tuple(map(int, np.mean(pixels, axis=0)))

        # K-means clustering to find dominant color
        criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 10, 1.0)
        k = min(3, len(colored_pixels))
        _, labels, centers = cv2.kmeans(colored_pixels, k, None, criteria,
                                        10, cv2.KMEANS_RANDOM_CENTERS)

        # Find most common cluster
        unique, counts = np.unique(labels, return_counts=True)
        dominant_idx = unique[np.argmax(counts)]
        dominant_color = centers[dominant_idx]

        # Convert BGR to RGB
        return (int(dominant_color[2]), int(dominant_color[1]), int(dominant_color[0]))

    def _is_background_color(self, color: Tuple[int, int, int]) -> bool:
        """Check if a color is likely a background color (white/gray)."""
        r, g, b = color

        # Check if grayscale-ish and light
        max_diff = max(abs(r-g), abs(g-b), abs(r-b))
        brightness = (r + g + b) / 3

        # White or very light gray
        if brightness > 230 and max_diff < 20:
            return True

        # Pure gray
        if max_diff < 10 and brightness > 180:
            return True

        return False

    def _rgb_to_hex(self, color: Tuple[int, int, int]) -> str:
        """Convert RGB tuple to hex string."""
        return "#{:02x}{:02x}{:02x}".format(color[0], color[1], color[2])

    def _extract_text(self, image: np.ndarray, events: List[CalendarEvent]) -> List[CalendarEvent]:
        """
        Extract text from each detected event block using OCR.
        """
        for event in events:
            x, y, w, h = int(event.x), int(event.y), int(event.width), int(event.height)

            # Extract block region with some padding
            pad = 2
            y1 = max(0, y - pad)
            y2 = min(image.shape[0], y + h + pad)
            x1 = max(0, x - pad)
            x2 = min(image.shape[1], x + w + pad)

            block_region = image[y1:y2, x1:x2]

            if block_region.size == 0:
                continue

            # Preprocess for OCR
            # Convert to grayscale
            gray = cv2.cvtColor(block_region, cv2.COLOR_BGR2GRAY)

            # Enhance contrast
            gray = cv2.convertScaleAbs(gray, alpha=1.5, beta=0)

            # Determine if text is light or dark
            mean_val = np.mean(gray)
            if mean_val < 128:
                # Light text on dark background - invert
                gray = cv2.bitwise_not(gray)

            # Resize if too small
            if h < 30:
                scale = 30 / h
                gray = cv2.resize(gray, None, fx=scale, fy=scale,
                                  interpolation=cv2.INTER_CUBIC)

            # OCR
            try:
                # Use PSM 6 for uniform block of text
                custom_config = r'--oem 3 --psm 6'
                text = pytesseract.image_to_string(gray, config=custom_config)
                event.text = text.strip().replace('\n', ' ')
            except Exception as e:
                if self.debug:
                    print(f"OCR failed for block at ({x}, {y}): {e}")

        return events

    def _normalize_coordinates(self, events: List[CalendarEvent],
                               width: int, height: int) -> List[CalendarEvent]:
        """Convert absolute pixel coordinates to relative (0-1) coordinates."""
        for event in events:
            event.x = event.x / width
            event.y = event.y / height
            event.width = event.width / width
            event.height = event.height / height
        return events

    def _filter_events(self, events: List[CalendarEvent]) -> List[CalendarEvent]:
        """
        Filter and validate detected events.

        - Remove duplicates
        - Remove overlapping detections (keep higher confidence)
        - Remove events with very low confidence
        """
        if not events:
            return events

        # Sort by confidence (highest first)
        events = sorted(events, key=lambda e: e.confidence, reverse=True)

        # Remove events with very low confidence
        events = [e for e in events if e.confidence > 0.3]

        # Remove overlapping events
        filtered = []
        for event in events:
            is_duplicate = False
            for existing in filtered:
                # Check IoU (Intersection over Union)
                iou = self._calculate_iou(event, existing)
                if iou > 0.5:
                    is_duplicate = True
                    break
            if not is_duplicate:
                filtered.append(event)

        return filtered

    def _calculate_iou(self, e1: CalendarEvent, e2: CalendarEvent) -> float:
        """Calculate Intersection over Union between two events."""
        x1 = max(e1.x, e2.x)
        y1 = max(e1.y, e2.y)
        x2 = min(e1.x + e1.width, e2.x + e2.width)
        y2 = min(e1.y + e1.height, e2.y + e2.height)

        if x1 >= x2 or y1 >= y2:
            return 0.0

        intersection = (x2 - x1) * (y2 - y1)
        area1 = e1.width * e1.height
        area2 = e2.width * e2.height
        union = area1 + area2 - intersection

        return intersection / union if union > 0 else 0.0

    def save_debug_images(self, output_dir: str):
        """Save debug images for visualization."""
        output_path = Path(output_dir)
        output_path.mkdir(parents=True, exist_ok=True)

        for name, img in self.debug_images.items():
            cv2.imwrite(str(output_path / f"debug_{name}.png"), img)


def analyze_calendar(image_path: str, output_path: str = None, debug: bool = False) -> dict:
    """
    Convenience function to analyze a calendar screenshot.

    Args:
        image_path: Path to the calendar screenshot
        output_path: Optional path to save JSON output
        debug: Enable debug mode for visualization

    Returns:
        Dictionary with analysis results
    """
    analyzer = CalendarAnalyzer(debug=debug)
    result = analyzer.analyze(image_path)

    output = {
        "imageWidth": result.image_width,
        "imageHeight": result.image_height,
        "calendar": result.calendar,
        "events": result.events
    }

    if output_path:
        with open(output_path, 'w', encoding='utf-8') as f:
            json.dump(output, f, indent=2, ensure_ascii=False)
        print(f"Analysis saved to: {output_path}")

    if debug:
        analyzer.save_debug_images(Path(image_path).parent / "debug")

    return output


def main():
    parser = argparse.ArgumentParser(
        description="Analyze calendar screenshot for 'Smash Your Week' game"
    )
    parser.add_argument("image", help="Path to calendar screenshot")
    parser.add_argument("-o", "--output", help="Output JSON file path")
    parser.add_argument("-d", "--debug", action="store_true",
                        help="Enable debug mode")
    parser.add_argument("-p", "--pretty", action="store_true",
                        help="Pretty print JSON to console")

    args = parser.parse_args()

    # Verify image exists
    if not Path(args.image).exists():
        print(f"Error: Image file not found: {args.image}")
        return 1

    # Set default output path
    output_path = args.output
    if not output_path:
        output_path = str(Path(args.image).with_suffix('.json'))

    try:
        result = analyze_calendar(args.image, output_path, debug=args.debug)

        if args.pretty:
            print("\n" + "="*50)
            print("ANALYSIS RESULT")
            print("="*50)
            print(json.dumps(result, indent=2, ensure_ascii=False))

        print(f"\nDetected {len(result['events'])} events")
        for i, event in enumerate(result['events'], 1):
            text_preview = event['text'][:30] + "..." if len(event['text']) > 30 else event['text']
            print(f"  {i}. {event['color']} - {text_preview or '(no text)'}")

        return 0

    except Exception as e:
        print(f"Error analyzing calendar: {e}")
        if args.debug:
            import traceback
            traceback.print_exc()
        return 1


if __name__ == "__main__":
    exit(main())

File: web/src/lib/analysis/calendar-analyzer.ts
/**
 * Calendar Screenshot Analyzer for "Smash Your Week" Game
 *
 * Browser-based version using Canvas API and Tesseract.js
 * Analyzes calendar screenshots and extracts event blocks.
 */

export interface CalendarEvent {
  x: number; // Relative X position (0-1)
  y: number; // Relative Y position (0-1)
  width: number; // Relative width (0-1)
  height: number; // Relative height (0-1)
  color: string; // Hex color code
  text: string; // Event text/title
  confidence: number; // Detection confidence (0-1)
  id: string; // Unique identifier
}

export interface CalendarGrid {
  days: string[];
  timeRange: {
    start: string;
    end: string;
  };
  columns: number;
  rows: number;
}

export interface AnalysisResult {
  imageWidth: number;
  imageHeight: number;
  imageDataUrl: string;
  calendar: CalendarGrid;
  events: CalendarEvent[];
}

interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}

interface ColorRGB {
  r: number;
  g: number;
  b: number;
}

/**
 * Main analyzer class for calendar screenshots
 */
export class CalendarAnalyzer {
  private canvas: HTMLCanvasElement;
  private ctx: CanvasRenderingContext2D;
  private imageData: ImageData | null = null;
  private width: number = 0;
  private height: number = 0;
  private enableOCR: boolean;

  // Configuration
  private readonly MIN_BLOCK_WIDTH_RATIO = 0.01; // Min width (1% of image)
  private readonly MIN_BLOCK_HEIGHT_RATIO = 0.005; // Min height (0.5% of image)
  private readonly MAX_BLOCK_WIDTH_RATIO = 0.5; // Max width (50% of image)
  private readonly MAX_BLOCK_HEIGHT_RATIO = 0.4; // Max height (40% of image)
  private readonly HEADER_RATIO = 0.05; // Top 5% is header/dates area - skip it
  private readonly MIN_ASPECT_RATIO = 0.2; // Min width/height ratio
  private readonly MAX_ASPECT_RATIO = 50; // Max width/height ratio

  constructor(enableOCR: boolean = false) {
    this.canvas = document.createElement("canvas");
    const ctx = this.canvas.getContext("2d", { willReadFrequently: true });
    if (!ctx) {
      throw new Error("Could not get canvas context");
    }
    this.ctx = ctx;
    this.enableOCR = enableOCR;
  }

  /**
   * Analyze a calendar screenshot
   */
  async analyze(imageSource: string | File | Blob): Promise<AnalysisResult> {
    // Load image
    const image = await this.loadImage(imageSource);
    this.width = image.width;
    this.height = image.height;

    // Set canvas size
    this.canvas.width = this.width;
    this.canvas.height = this.height;

    // Draw image to canvas
    this.ctx.drawImage(image, 0, 0);
    this.imageData = this.ctx.getImageData(0, 0, this.width, this.height);

    // Get image as data URL for game background
    const imageDataUrl = this.canvas.toDataURL("image/png");

    // Detect calendar structure
    const calendar = this.detectGrid();

    // Detect event blocks
    let events = this.detectEvents();

    // Extract text if OCR is enabled
    if (this.enableOCR) {
      events = await this.extractTextOCR(events);
    }

    // Normalize coordinates to relative values
    events = this.normalizeCoordinates(events);

    // Filter overlapping/duplicate events
    events = this.filterEvents(events);

    // Add unique IDs
    events = events.map((event, index) => ({
      ...event,
      id: `event-${index}-${Date.now()}`,
    }));

    return {
      imageWidth: this.width,
      imageHeight: this.height,
      imageDataUrl,
      calendar,
      events,
    };
  }

  /**
   * Load image from various sources
   */
  private loadImage(source: string | File | Blob): Promise<HTMLImageElement> {
    return new Promise((resolve, reject) => {
      const img = new Image();
      img.crossOrigin = "anonymous";

      img.onload = () => resolve(img);
      img.onerror = () => reject(new Error("Failed to load image"));

      if (typeof source === "string") {
        img.src = source;
      } else {
        img.src = URL.createObjectURL(source);
      }
    });
  }

  /**
   * Detect calendar grid structure
   */
  private detectGrid(): CalendarGrid {
    // Default structure - can be enhanced with more sophisticated detection
    return {
      days: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
      timeRange: {
        start: "08:00",
        end: "18:00",
      },
      columns: 7,
      rows: 24,
    };
  }

  /**
   * Detect colored event blocks in the calendar
   */
  private detectEvents(): CalendarEvent[] {
    if (!this.imageData) return [];

    const events: CalendarEvent[] = [];
    const visited = new Set<string>();
    const data = this.imageData.data;

    // Skip the header area (dates at the top)
    const startY = Math.floor(this.height * this.HEADER_RATIO);

    // Scan the image for colored regions (skip header)
    for (let y = startY; y < this.height; y += 2) {
      for (let x = 0; x < this.width; x += 2) {
        const key = `${x},${y}`;
        if (visited.has(key)) continue;

        const idx = (y * this.width + x) * 4;
        const r = data[idx];
        const g = data[idx + 1];
        const b = data[idx + 2];

        // Check if this pixel is a colored block (not background)
        if (this.isColoredPixel(r, g, b)) {
          // Flood fill to find block boundaries
          const rect = this.floodFillBounds(x, y, visited);

          if (rect && this.isValidBlockSize(rect)) {
            const color = this.getDominantColor(rect);
            events.push({
              x: rect.x,
              y: rect.y,
              width: rect.width,
              height: rect.height,
              color: this.rgbToHex(color),
              text: "",
              confidence: this.calculateConfidence(rect),
              id: "",
            });
          }
        }
      }
    }

    return events;
  }

  /**
   * Check if a pixel is part of a colored block (not background)
   * A colored block is: any rectangle with color that is NOT:
   * - A grey calendar grid line (R ≈ G ≈ B)
   * - A white/near-white area
   */
  private isColoredPixel(r: number, g: number, b: number): boolean {
    // Check brightness
    const brightness = (r + g + b) / 3;

    // Filter out white/near-white pixels (background)
    if (brightness > 250) return false;

    // Check if pixel is grey (calendar grid lines have R ≈ G ≈ B)
    const maxDiff = Math.max(
      Math.abs(r - g),
      Math.abs(g - b),
      Math.abs(r - b)
    );

    // Only filter truly grey pixels (very similar R, G, B values)
    // Colored blocks have at least some difference between channels
    const isGrey = maxDiff < 10;

    // Filter out grey pixels (grid lines, borders)
    if (isGrey) return false;

    // Accept any pixel that has color (not grey, not white)
    return true;
  }

  /**
   * Flood fill to find block boundaries
   */
  private floodFillBounds(
    startX: number,
    startY: number,
    visited: Set<string>
  ): Rect | null {
    if (!this.imageData) return null;

    const data = this.imageData.data;
    const startIdx = (startY * this.width + startX) * 4;
    const startColor = {
      r: data[startIdx],
      g: data[startIdx + 1],
      b: data[startIdx + 2],
    };

    // BFS to find connected colored region
    const queue: [number, number][] = [[startX, startY]];
    let minX = startX,
      maxX = startX;
    let minY = startY,
      maxY = startY;
    let pixelCount = 0;
    const maxPixels = 50000; // Limit to prevent infinite loops

    while (queue.length > 0 && pixelCount < maxPixels) {
      const [x, y] = queue.shift()!;
      const key = `${x},${y}`;

      if (
        visited.has(key) ||
        x < 0 ||
        x >= this.width ||
        y < 0 ||
        y >= this.height
      ) {
        continue;
      }

      const idx = (y * this.width + x) * 4;
      const r = data[idx];
      const g = data[idx + 1];
      const b = data[idx + 2];

      // Check if pixel is similar to start color
      if (!this.isColorSimilar(startColor, { r, g, b }, 50)) {
        continue;
      }

      visited.add(key);
      pixelCount++;

      // Update bounds
      minX = Math.min(minX, x);
      maxX = Math.max(maxX, x);
      minY = Math.min(minY, y);
      maxY = Math.max(maxY, y);

      // Add neighbors (4-connectivity with step)
      const step = 2;
      queue.push([x + step, y], [x - step, y], [x, y + step], [x, y - step]);
    }

    if (pixelCount < 15) return null; // Too small - filter noise

    return {
      x: minX,
      y: minY,
      width: maxX - minX + 1,
      height: maxY - minY + 1,
    };
  }

  /**
   * Check if two colors are similar
   */
  private isColorSimilar(
    c1: ColorRGB,
    c2: ColorRGB,
    threshold: number
  ): boolean {
    const dr = Math.abs(c1.r - c2.r);
    const dg = Math.abs(c1.g - c2.g);
    const db = Math.abs(c1.b - c2.b);
    return dr + dg + db < threshold;
  }

  /**
   * Check if block size is valid
   */
  private isValidBlockSize(rect: Rect): boolean {
    const widthRatio = rect.width / this.width;
    const heightRatio = rect.height / this.height;
    const aspectRatio = rect.width / rect.height;

    // Check size constraints
    const validWidth = widthRatio >= this.MIN_BLOCK_WIDTH_RATIO &&
                       widthRatio <= this.MAX_BLOCK_WIDTH_RATIO;
    const validHeight = heightRatio >= this.MIN_BLOCK_HEIGHT_RATIO &&
                        heightRatio <= this.MAX_BLOCK_HEIGHT_RATIO;

    // Check aspect ratio to filter out lines (grid lines, dividers)
    const validAspect = aspectRatio >= this.MIN_ASPECT_RATIO &&
                        aspectRatio <= this.MAX_ASPECT_RATIO;

    return validWidth && validHeight && validAspect;
  }

  /**
   * Get dominant color of a region
   */
  private getDominantColor(rect: Rect): ColorRGB {
    if (!this.imageData) return { r: 128, g: 128, b: 128 };

    const data = this.imageData.data;
    const colorCounts = new Map<string, { color: ColorRGB; count: number }>();

    // Sample pixels in the region
    const sampleStep = Math.max(1, Math.floor(rect.width / 20));

    for (let y = rect.y; y < rect.y + rect.height; y += sampleStep) {
      for (let x = rect.x; x < rect.x + rect.width; x += sampleStep) {
        if (x >= this.width || y >= this.height) continue;

        const idx = (y * this.width + x) * 4;
        const r = data[idx];
        const g = data[idx + 1];
        const b = data[idx + 2];

        // Quantize color for grouping
        const qr = Math.round(r / 20) * 20;
        const qg = Math.round(g / 20) * 20;
        const qb = Math.round(b / 20) * 20;
        const key = `${qr},${qg},${qb}`;

        const existing = colorCounts.get(key);
        if (existing) {
          existing.count++;
        } else {
          colorCounts.set(key, { color: { r: qr, g: qg, b: qb }, count: 1 });
        }
      }
    }

    // Find most common non-white color
    let maxCount = 0;
    let dominantColor: ColorRGB = { r: 128, g: 128, b: 128 };

    for (const { color, count } of colorCounts.values()) {
      // Skip near-white colors
      if (color.r > 230 && color.g > 230 && color.b > 230) continue;

      if (count > maxCount) {
        maxCount = count;
        dominantColor = color;
      }
    }

    return dominantColor;
  }

  /**
   * Calculate detection confidence
   */
  private calculateConfidence(rect: Rect): number {
    const area = rect.width * rect.height;
    const imageArea = this.width * this.height;
    const areaRatio = area / imageArea;

    // Confidence based on size (reasonable event blocks)
    if (areaRatio > 0.001 && areaRatio < 0.1) {
      return 0.8;
    } else if (areaRatio > 0.0005 && areaRatio < 0.15) {
      return 0.6;
    }
    return 0.4;
  }

  /**
   * Extract text using OCR (Tesseract.js)
   * Note: OCR is disabled by default for performance. Enable if needed.
   */
  private async extractTextOCR(
    events: CalendarEvent[]
  ): Promise<CalendarEvent[]> {
    // OCR disabled - would require tesseract.js dependency
    // To enable: npm install tesseract.js and uncomment below
    console.log("OCR disabled - blocks will not have text labels");
    return events;
  }

  /**
   * Normalize coordinates to relative values (0-1)
   */
  private normalizeCoordinates(events: CalendarEvent[]): CalendarEvent[] {
    return events.map((event) => ({
      ...event,
      x: event.x / this.width,
      y: event.y / this.height,
      width: event.width / this.width,
      height: event.height / this.height,
    }));
  }

  /**
   * Filter overlapping and duplicate events
   */
  private filterEvents(events: CalendarEvent[]): CalendarEvent[] {
    // Sort by confidence
    events.sort((a, b) => b.confidence - a.confidence);

    // Filter low confidence
    events = events.filter((e) => e.confidence > 0.3);

    // Remove overlapping
    const filtered: CalendarEvent[] = [];
    for (const event of events) {
      let isOverlapping = false;
      for (const existing of filtered) {
        const iou = this.calculateIoU(event, existing);
        if (iou > 0.3) {
          isOverlapping = true;
          break;
        }
      }
      if (!isOverlapping) {
        filtered.push(event);
      }
    }

    return filtered;
  }

  /**
   * Calculate Intersection over Union
   */
  private calculateIoU(e1: CalendarEvent, e2: CalendarEvent): number {
    const x1 = Math.max(e1.x, e2.x);
    const y1 = Math.max(e1.y, e2.y);
    const x2 = Math.min(e1.x + e1.width, e2.x + e2.width);
    const y2 = Math.min(e1.y + e1.height, e2.y + e2.height);

    if (x1 >= x2 || y1 >= y2) return 0;

    const intersection = (x2 - x1) * (y2 - y1);
    const area1 = e1.width * e1.height;
    const area2 = e2.width * e2.height;
    const union = area1 + area2 - intersection;

    return union > 0 ? intersection / union : 0;
  }

  /**
   * Convert RGB to hex
   */
  private rgbToHex(color: ColorRGB): string {
    return (
      "#" +
      [color.r, color.g, color.b]
        .map((c) => Math.max(0, Math.min(255, c)).toString(16).padStart(2, "0"))
        .join("")
    );
  }

  /**
   * Convert RGB to HSL
   */
  private rgbToHsl(
    r: number,
    g: number,
    b: number
  ): { h: number; s: number; l: number } {
    r /= 255;
    g /= 255;
    b /= 255;

    const max = Math.max(r, g, b);
    const min = Math.min(r, g, b);
    let h = 0;
    let s = 0;
    const l = ((max + min) / 2) * 100;

    if (max !== min) {
      const d = max - min;
      s = (l > 50 ? d / (2 - max - min) : d / (max + min)) * 100;

      switch (max) {
        case r:
          h = ((g - b) / d + (g < b ? 6 : 0)) / 6;
          break;
        case g:
          h = ((b - r) / d + 2) / 6;
          break;
        case b:
          h = ((r - g) / d + 4) / 6;
          break;
      }
      h *= 360;
    }

    return { h, s, l };
  }
}

/**
 * Convenience function to analyze a calendar image
 */
export async function analyzeCalendar(
  imageSource: string | File | Blob,
  enableOCR: boolean = false
): Promise<AnalysisResult> {
  const analyzer = new CalendarAnalyzer(enableOCR);
  return analyzer.analyze(imageSource);
}

export default CalendarAnalyzer;
